tmt/steps/provision/testcloud.py

```diff
@@ -219,6 +219,19 @@ class GuestTestcloud(tmt.GuestSsh):
     _image: Optional['testcloud.image.Image'] = None  # type: ignore[name-defined]
     _instance: Optional['testcloud.instance.Instance'] = None  # type: ignore[name-defined]
 
+    @property
+    def is_ready(self) -> bool:
+        if self._instance is None:
+            return False
+
+        assert testcloud is not None
+        assert libvirt is not None
+        try:
+            state = testcloud.instance._find_domain(self._instance.name, self._instance.connection)
+            return bool(state == 'running')
```

Comment:
Why is the explicit `bool` needed here?

Reply:
Hi @psss, it is because of `mypy` checking. e.g.

```patch
$ git diff
diff --git a/tmt/steps/provision/testcloud.py b/tmt/steps/provision/testcloud.py
index c404c7a..1b84757 100644
--- a/tmt/steps/provision/testcloud.py
+++ b/tmt/steps/provision/testcloud.py
@@ -228,7 +228,7 @@ class GuestTestcloud(tmt.GuestSsh):
         assert libvirt is not None
         try:
             state = testcloud.instance._find_domain(self._instance.name, self._instance.connection)
-            return bool(state == 'running')
+            return state == 'running'
         except libvirt.libvirtError:
             return False
```


```bash
$ git commit -s -a -m "Remove bool()"
...<snip>...
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tmt/steps/provision/testcloud.py:231: error: Returning Any from function declared to return "bool"  [no-any-return]
Found 1 error in 1 file (checked 40 source files)
```

Any other suggestion to avoid such an error?

Reply:
Can you check what's the type of `state`? `reveal_type(state)` & `pre-commit run --all-files mypy` should report it.

This happens often with libraries that are not annotated, which may mean the return value type of `_find_domain()` would be `Any`, and `Any == bool` is not guaranteed to be `bool` as well, it may be `Any`thing.

There are two basic ways how to deal with this from our end:

* `state = cast(str, testcloud.instance._find_domain(self._instance.name, self._instance.connection))` - `state` would have a narrower type, mypy would be happy. This requires checking the library method, finding out what values and types it may return, maybe the right value is `Optional[str]`, or even `Union[None, str, int, float]`. Better when the return value is used in more complex code, e.g. printed, transformed, passed to other parts of code - it would have known type when crossing the border between a library and tmt code, which is the right place.
* `bool(state == 'running')` - `state` is not used anywhere outside of this scope, never passed elsewhere, it's used only for comparison against a string. And we're interested in checking whether the state is "running", i.e. everything else (`1`, `"foo"`, `None`, ...) means `False` to us anyway. I'd pick this solution, but a comment might help future engineers ("_find_domain() returns Any, make sure we return a boolean & make mypy happy")
* `cast(bool, state == 'running')` would be a worse choice, as we don't really know what the value is, and pretending it's a boolean could introduce bugs down the stream (probably not here, but it's a bad idea in general - if we *knew* `state` is a string, it'd be better to apply `cast()` when `_find_domain` is called, i.e. the first option.

Reply:
> and `Any == bool` is not guaranteed to be `bool` as well, it may be `Any`thing.

Hmmm, this is interesting. Would not expect that comparison can return anything else than `bool`...

Reply:
From the suggested solutions I'd say the second one probably makes sense the most.

Reply:
> > and `Any == bool` is not guaranteed to be `bool` as well, it may be `Any`thing.
> 
> Hmmm, this is interesting. Would not expect that comparison can return anything else than `bool`...

Well, you're in for a surprise, custom operators can introduce new levels of pain :)

```
$ cat t.py
class Foo:
  def __eq__(self, other):
    if other is self:
        return 'same!'
    return 'not same!'

bar = Foo()
baz = Foo()

print(bar == baz)
print(bar == bar)

print(bool(bar == baz))
print(bool(bar == bar))
$ python t.py
not same!
same!
True
True
```

Reply:
> Can you check what's the type of state? reveal_type(state) & pre-commit run --all-files mypy should report it.

@happz, the type of `state` is `Any`. e.g.
```patch
--- a/tmt/steps/provision/testcloud.py
+++ b/tmt/steps/provision/testcloud.py
@@ -228,7 +228,8 @@ class GuestTestcloud(tmt.GuestSsh):
         assert libvirt is not None
         try:
             state = testcloud.instance._find_domain(self._instance.name, self._instance.connection)
-            return bool(state == 'running')
+            reveal_type(state)
+            return state == 'running'
         except libvirt.libvirtError:
             return False
``` 

```bash
$ pre-commit run --all-files mypy
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

tmt/steps/provision/testcloud.py:231: note: Revealed type is "Any"
tmt/steps/provision/testcloud.py:232: error: Returning Any from function declared to return "bool"  [no-any-return]
Found 1 error in 1 file (checked 40 source files)
```

> From the suggested solutions I'd say the second one probably makes sense the most.

@psss, I have updated the related line, please take a look again, thanks!

Reply:
Oh, yeah, language flexibility is always nice! :)